### ISSUE TITLE
The Pytorch model trained by yolov8-seg on its own data works well, but it does not work well after switching to the onnx model, and the images and parameters remain the same

### ISSUE THREAD

**Issue by qingqiqiu** [ultralytics/ultralytics#14597]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I used the YOLOv8 segmentation model to train on my own data, the pytorch model works very well, but the effect is very poor after switching to the ONNX model, the official model doesn't seem to have this This has anyone encountered this problem? How to solve it? ...

**Comment by Y-T-G** [comment]
Are you using ONNX directly in ultralytics?

**Comment by qingqiqiu** [comment]
yes，I trained my data with yolov10 repository which included yolov8，and I tryed exported to onnx with yolov10 repository and ultralytics repository and tryed different environment，but the pytorch model performed very well, can't get good result with onnx model. The model and the result image is ttached, thanks for your help!






















At 2024-07-22 19:52:28, "Mohammed Yasin" ***@***.***> wrote:

Are you using ONNX directly in ultralytics?

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***>

从网易163邮箱发来的超大附件推荐客户端极速下载
image0-pytorch.jpg (931.43K, 2024年8月7日 10:53 到期)
下载
image0-onnx.jpg (792.04K, 2024年8月7日 10:53 到期)
下载
20231221153206_5600411_0.png (3.72M, 2024年8月7日 10:53 到期)
下载
yolov8_seg_1_last.pt (52.34M, 2024年8月7日 10:53 到期)
下载
yolov8_seg_1_last.onnx (104.04M, 2024年8月7日 10:53 到期)
下载

**Comment by qingqiqiu** [comment]
> Are you using ONNX directly in ultralytics?

yes，I trained my data with yolov10 repository which included yolov8，and I tryed exported to onnx with yolov10 repository and ultralytics repository and tryed different environment，but the pytorch model performed very well, can't get good result with onnx model.

**Comment by pderrenger** [comment]
@qingqiqiu thank you for sharing the details and attachments. It seems like the ONNX model is not performing as expected compared to the PyTorch model. This discrepancy can sometimes occur due to differences in how the models handle certain operations or due to export issues.

First, ensure you are using the latest versions of the Ultralytics and ONNX packages. If the issue persists, try the following steps:
1. Verify that the ONNX model is exported with the same input size and parameters as the PyTorch model.
2. Check if there are any warnings or errors during the export process.
3. Validate the ONNX model using the ONNX runtime to ensure it produces consistent outputs.

If the problem continues, please provide more details about the environment and any specific errors or warnings you encounter. This will help us diagnose the issue more effectively.

**Comment by qingqiqiu** [comment]
> @qingqiqiu thank you for sharing the details and attachments. It seems like the ONNX model is not performing as expected compared to the PyTorch model. This discrepancy can sometimes occur due to differences in how the models handle certain operations or due to export issues.
> 
> First, ensure you are using the latest versions of the Ultralytics and ONNX packages. If the issue persists, try the following steps:
> 
> 1. Verify that the ONNX model is exported with the same input size and parameters as the PyTorch model.
> 2. Check if there are any warnings or errors during the export process.
> 3. Validate the ONNX model using the ONNX runtime to ensure it produces consistent outputs.
> 
> If the problem continues, please provide more details about the environment and any specific errors or warnings you encounter. This will help us diagnose the issue more effectively.

![image](https://github.com/user-attachments/assets/a6cf2405-0150-494e-809e-2ac2a1d412ca)
I trained, export, test the model with the same input size 640.
There are no warnings or errors during the export process.
And I used the source code with AutoBackend to load and setup onnx model, which used onnxruntime

My onnx version is :
 pip list|grep onnx                                                                                 │  Note that the developm
onnx                     1.15.0                                                                                                                                             │ent build is not optimiz
onnxruntime              1.15.1                                                                                                                                             │ed.
onnxruntime-gpu          1.15.1

**Comment by pderrenger** [comment]
@qingqiqiu thank you for the detailed information. Given that the official model works well in ONNX format, the issue might be specific to your custom model. Here are a few additional steps to consider:

1. Ensure that the preprocessing steps (e.g., normalization, resizing) are consistent between the PyTorch and ONNX models.
2. Check if there are any differences in the post-processing steps, such as Non-Maximum Suppression (NMS), between the two models.
3. Compare the outputs of the PyTorch and ONNX models layer by layer to identify where the discrepancy begins.

If the issue persists, please provide more details about the specific differences in performance between the PyTorch and ONNX models. This will help us further diagnose the problem.

**Comment by qingqiqiu** [comment]
> @qingqiqiu thank you for the detailed information. Given that the official model works well in ONNX format, the issue might be specific to your custom model. Here are a few additional steps to consider:
> 
> 1. Ensure that the preprocessing steps (e.g., normalization, resizing) are consistent between the PyTorch and ONNX models.
> 2. Check if there are any differences in the post-processing steps, such as Non-Maximum Suppression (NMS), between the two models.
> 3. Compare the outputs of the PyTorch and ONNX models layer by layer to identify where the discrepancy begins.
> 
> If the issue persists, please provide more details about the specific differences in performance between the PyTorch and ONNX models. This will help us further diagnose the problem.
![yolov8m-seg-pt](https://github.com/user-attachments/assets/25bbeb69-b014-464e-8c3a-a38b8a2c622f)
![yolov8m-seg-onnx](https://github.com/user-attachments/assets/6a705a88-6059-4736-a0a1-bed039d4e770)
Thanks for your answer. I run yolov8m-seg.pt and yolov8m-seg.onnx model in bus.jpg with same params and settings, and found that the result is different, pytorch model result is beter  than onnx model. onnx model run the image with 640*640 shape, but pytorch model test the image with shape 640*480, will this influence the results? And the test running log is:
~/workspace/training_tools/iliad/scripts/yolov10$ python test.py --type yolov8_seg --weights /home/pikachu/workspace/training_tools/iliad/scripts/yolov10/yolov8m-seg.pt  --image ./ultralytics/assets/bus.jpg 
Namespace(conf=0.25, device='cpu', image='./ultralytics/assets/bus.jpg', type='yolov8_seg', weights='/home/pikachu/workspace/training_tools/iliad/scripts/yolov10/yolov8m-seg.pt') opt
Ultralytics YOLOv8.1.34 🚀 Python-3.8.12 torch-2.0.1+cu117 CPU (AMD EPYC 7542 32-Core Processor)
YOLOv8m-seg summary (fused): 245 layers, 27268704 parameters, 0 gradients, 110.2 GFLOPs

<class 'list'> im.shape
<class 'numpy.ndarray'> i.shape
(1080, 810, 3)
True not_tensor
False self.model.fp16
<class 'torch.Tensor'> torch.Size([1, 3, 640, 480]) im.shape
0: 640x480 4 persons, 1 bus, 1 tie, 392.5ms
Speed: 7.4ms preprocess, 392.5ms inference, 11.6ms postprocess per image at shape (1, 3, 640, 480)
Results saved to /home/pikachu/workspace/yolov10/runs/segment/predict22
predict time 0.45508766174316406

pikachu@Pikachu:~/workspace/training_tools/iliad/scripts/yolov10$ python test.py --type yolov8_seg --weights /home/pikachu/workspace/training_tools/iliad/scripts/yolov10/yolov8m-seg.onnx  --image ./ultralytics/assets/bus.jpg 
Namespace(conf=0.25, device='cpu', image='./ultralytics/assets/bus.jpg', type='yolov8_seg', weights='/home/pikachu/workspace/training_tools/iliad/scripts/yolov10/yolov8m-seg.onnx') opt
Ultralytics YOLOv8.1.34 🚀 Python-3.8.12 torch-2.0.1+cu117 CPU (AMD EPYC 7542 32-Core Processor)
Loading /home/pikachu/workspace/training_tools/iliad/scripts/yolov10/yolov8m-seg.onnx for ONNX Runtime inference...
<class 'list'> im.shape
<class 'numpy.ndarray'> i.shape
(1080, 810, 3)
True not_tensor
False self.model.fp16
<class 'torch.Tensor'> torch.Size([1, 3, 640, 640]) im.shape
0: 640x640 4 persons, 1 bus, 1 tie, 549.6ms
Speed: 26.8ms preprocess, 549.6ms inference, 25.0ms postprocess per image at shape (1, 3, 640, 640)
Results saved to /home/pikachu/workspace/yolov10/runs/segment/predict23
predict time 0.6453349590301514

**Comment by TClan8023** [comment]
hi, have you solved the problems here? I found the same bug.

**Comment by Y-T-G** [comment]
What's the output of `yolo checks`?

**Comment by qingqiqiu** [comment]
> hi, have you solved the problems here? I found the same bug.

sorry, I didn't solve this problem, I downloaded a onnx file, and compared with my onnx file which is exported with same torch and onnx version，but the model struction have many difference. 
![image](https://github.com/user-attachments/assets/5f936a84-bed6-4766-b329-f14b87cd7ec5)

**Comment by qingqiqiu** [comment]
> yolo checks

thanks for your answer! my yolo checks output is:
![image](https://github.com/user-attachments/assets/d89010a1-1255-4d0d-bbee-4e22bc350993)

**Comment by Y-T-G** [comment]
Try torch `2.3.0`

**Comment by qingqiqiu** [comment]
> > yolo checks
> 
> thanks for your answer! my yolo checks output is: ![image](https://private-user-images.githubusercontent.com/26505942/359348912-d89010a1-1255-4d0d-bbee-4e22bc350993.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQxMzI0NTcsIm5iZiI6MTcyNDEzMjE1NywicGF0aCI6Ii8yNjUwNTk0Mi8zNTkzNDg5MTItZDg5MDEwYTEtMTI1NS00ZDBkLWJiZWUtNGUyMmJjMzUwOTkzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIwVDA1MzU1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJhZmFkNjEwNWZiMTQ4Yzg0YjFiZDMwMmQ2Y2QzOWQ0OWVhMTc2YjljYzhmMDdiNjBlNjE0MjFkZDE4ODI1NmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.YQTUSYG6VPAX38YNTxDGaFX59Di1AaT48cz9Ax2URtg)
I also tried different version of torch, above is yolo checks outputs for torch 0.12.0, and this is yolo checks outputs for torch=2.0.1
![image](https://github.com/user-attachments/assets/7bb0de1e-83c4-4547-bd55-1a43cae50cec)

**Comment by Y-T-G** [comment]
What's the `imgsz` of your model?

**Comment by qingqiqiu** [comment]
> What's the `imgsz` of your model?
I used default imgsz 640

**Comment by qingqiqiu** [comment]
> Try torch `2.3.0`

I tried torch 2.3.0, but the exported onnx result is still not as good as .pt model and the downloaded onnx model

**Comment by Y-T-G** [comment]
You're using Ultralytics 8.1.34 it is probably the YOLOv10 fork and not the the official ultralytics.

Install the official ultralytics. `pip install -U ultralytics `

**Comment by qingqiqiu** [comment]
> You're using Ultralytics 8.1.34 it is probably the YOLOv10 fork and not the the official ultralytics.
> 
> Install the official ultralytics. `pip install -U ultralytics `

I tried with official ultralytics with old version but it not worked. Now I pull the code and tried again, it worked well. Thank you very much!!

**Comment by qingqiqiu** [comment]
> hi, have you solved the problems here? I found the same bug.

I took the suggestion of   @Y-T-G [Y-T-G](https://github.com/Y-T-G) and updated the official library code to solve this problem, it is fixed.
